cfme/common/vm.py

```diff
@@ -651,6 +657,143 @@ def does_vm_exist_on_provider(self):
+                BaseVM.STATE_ON, BaseVM.STATE_OFF, BaseVM.STATE_SUSPENDED, BaseVM.STATE_PAUSED
+        """
+        def _return_false():
+            """Used when a state check isn't needed"""
+            return False
+
+        def _needs_prep_to_be_stopped():
+            """Check if VM is suspended/paused via provider API."""
+            return self.is_suspended_on_provider() or self.is_paused_on_provider()
+
+        def _needs_prep_to_be_suspended():
+            """Check if VM is stopped/paused via provider API."""
+            return self.is_stopped_on_provider() or self.is_paused_on_provider()
+
+        def _needs_prep_to_be_paused():
+            """Check if VM is stopped/suspended via provider API."""
+            return self.is_stopped_on_provider() or self.is_suspended_on_provider()
+
+        def _able_to_start(self):
+            """Check if VM is stopped, suspended, or paused via provider API."""
+            return (self.is_stopped_on_provider() or
+                    self.is_suspended_on_provider() or
+                    self.is_paused_on_provider())
+
+        # Note: each of the vars defined below are pointing *to methods*
+        if state == self.RUNNING:
+            in_desired_state = self.is_running_on_provider
+            in_state_requiring_prep = _return_false
+            in_actionable_state = _able_to_start
+            do_prep = None  # do_prep will never be needed for this state
+            do_action = self.start_on_provider
+
+        elif state == self.STOPPED:
+            in_desired_state = self.is_stopped_on_provider
+            in_state_requiring_prep = _needs_prep_to_be_stopped
+            in_actionable_state = self.is_running_on_provider
+            do_prep = self.start_on_provider
+            do_action = self.stop_on_provider
+
+        elif state == self.SUSPENDED:
+            in_desired_state = self.is_suspended_on_provider
+            in_state_requiring_prep = _needs_prep_to_be_suspended
+            in_actionable_state = self.is_running_on_provider
+            do_prep = self.start_on_provider
+            do_action = self.suspend_on_provider
+
+        elif state == self.PAUSED:
+            in_desired_state = self.is_paused_on_provider
+            in_state_requiring_prep = _needs_prep_to_be_paused
+            in_actionable_state = self.is_running_on_provider
+            do_prep = self.start_on_provider
+            do_action = self.pause_on_provider
+
+        else:
+            raise ValueError("Invalid state '{}'".format(state))
+
+        @wait_for_decorator(timeout="6m", delay=15, message="vm to reach state '{}'".format(state))
+        def vm_reached_desired_state():
+            if in_desired_state():
+                return True
+            elif in_state_requiring_prep():
+                do_prep()
+            elif in_actionable_state():
+                do_action()
+
+            logger.debug(
+                "Sleeping 15secs...(current state: {}, needed state: {})".format(
```

Comment:
REQUIRED: hard coded value and typo

Reply:
timeout/delay are no longer hard-coded... but what is the typo, am I blind? :P

Reply:
by typo I meant 15secs :)